engine/engine.go

```diff
@@ -0,0 +1,132 @@
+	// is used for feature checking. Though most methods only needs to return
+	// ErrFeatureNotSupported when called, rather than being declared here.
+	FeatureSet() FeatureSet
+	// NewSandboxBuilder returns a new instance of the SandboxBuilder interface.
+	//
+	// We'll create a SandboxBuilder for each task run. This is really a setup
+	// step where the implementor may acquire resources referenced in the
+	// SandboxOptions.
+	//
+	// Example: An engine implementation based on docker, may start downloading
+	// the docker image in before returning from NewSandboxBuilder(). The
+	// SandboxBuilder instance returned will then reference the docker image
+	// downloading process, and be ready to start a new docker container once
+	// StartSandbox() is called, Obviously blocking that call until docker image
+	// download is completed.
+	//
+	// This operation should parse the engine-specific payload parts given in
+	// SandboxOptions and return a MalformedPayloadError error if the payload
+	// is invalid.
+	//
+	// Non-fatal errors: MalformedPayloadError, ErrEngineIsSingleton.
+	NewSandboxBuilder(options *SandboxOptions, context *runtime.SandboxContext) (SandboxBuilder, error)
+	// NewCacheFolder returns a new Volume backed by a file system folder
+	// if cache-folders folders are supported, otherwise it must return
+	// ErrFeatureNotSupported.
+	//
+	// Non-fatal errors: ErrFeatureNotSupported
+	NewCacheFolder() (Volume, error)
+	// NewMemoryDisk returns a new Volume backed by a ramdisk, if ramdisks are
+	// supported, otherwise it must return ErrFeatureNotSupported.
+	//
+	// Non-fatal errors: ErrFeatureNotSupported
+	NewMemoryDisk() (Volume, error)
+}
+
+// The FeatureSet structure defines the set of features supported by an engine.
+//
+// Some plugins will use this for feature detection, most plugins will call the
+// methods in question and handle the ErrFeatureNotSupported error. For this
+// reason it's essential to also return ErrFeatureNotSupported from methods
+// related to unsupported features (see docs of individual methods).
+//
+// Plugin implementors are advised to call methods and handling unsupported
+// features by handling ErrFeatureNotSupported errors. But in some cases it
+// might be necessary to adjust behavior in case of unsupported methods, for
+// this upfront feature checking using FeatureSet is necessary.
+//
+// To encourage the try and handle errors pattern, the FeatureSet shall only
+// list features for which we critically need upfront feature testing.
+type FeatureSet struct {
+	IsSingletonEngine bool
+}
+
+// NewEngine returns an engine implementation, if the requested engine is
+// available under the current build contraints, GOOS, GOARCH.
+//
+// This function is intended to be called once immediately after startup, with
+// engine of choice as given by configuration.
+func NewEngine(engineName string, runtime *runtime.EngineContext) (Engine, error) {
```

Comment:
Should we call this simply `New` instead of `NewEngine`? See _Simplify function names_ in the [go blog](https://blog.golang.org/package-names).